File: backend/scripts/smoke.ts
import { spawn } from "node:child_process";
import { setTimeout as delay } from "node:timers/promises";

const HEALTH_URL = "http://127.0.0.1:8000/health";
const TIMEOUT_MS = 30_000;

async function waitForHealth(): Promise<boolean> {
  const deadline = Date.now() + TIMEOUT_MS;
  while (Date.now() < deadline) {
    try {
      const response = await fetch(HEALTH_URL);
      if (response.ok) {
        return true;
      }
    } catch {
      // the server is not listening yet
    }
    await delay(500);
  }
  return false;
}

const server = spawn("node", ["dist/main.js"], { stdio: "inherit" });

try {
  if (await waitForHealth()) {
    console.log(`OK: ${HEALTH_URL} answered 200`);
  } else {
    console.error(`FAIL: ${HEALTH_URL} never answered within ${TIMEOUT_MS}ms`);
    process.exitCode = 1;
  }
} finally {
  server.kill("SIGTERM");
}

File: backend/scripts/smoke.py
import sys
import time
import urllib.error
import urllib.request
from contextlib import suppress
from subprocess import Popen

HEALTH_URL = "http://127.0.0.1:8000/health"
TIMEOUT_SECONDS = 30


def wait_for_health() -> bool:
    deadline = time.monotonic() + TIMEOUT_SECONDS
    while time.monotonic() < deadline:
        with (
            suppress(urllib.error.URLError, ConnectionError),
            urllib.request.urlopen(HEALTH_URL, timeout=2) as response,
        ):
            return bool(response.status == 200)
        time.sleep(0.5)
    return False


def main() -> int:
    server = Popen([sys.executable, "-m", "uvicorn", "sispaniers.main:app", "--port", "8000"])
    try:
        if not wait_for_health():
            print(f"FAIL: {HEALTH_URL} never answered within {TIMEOUT_SECONDS}s")
            return 1
        print(f"OK: {HEALTH_URL} answered 200")
        return 0
    finally:
        server.terminate()
        server.wait(timeout=10)


if __name__ == "__main__":
    raise SystemExit(main())
